crates/ide-assists/src/handlers/recursively_add_derive.rs

```diff
@@ -0,0 +1,1010 @@
+            .any(|macro_| self.derive_macro == macro_)
+            // If it does not, and there is a trait which the derive (likely) implements, check if the type has already implemented it.
+            || matches!(self.maybe_trait, Some(trait_) if adt.ty(ctx.db()).impls_trait(ctx.db(), trait_, &[]))
+    }
+}
+
+impl From<DeriveItem> for ItemInNs {
+    fn from(value: DeriveItem) -> Self {
+        value.derive_macro.into()
+    }
+}
+
+fn field_list_adts<'a>(
+    ctx: &'a AssistContext<'_>,
+    list: ast::FieldList,
+) -> impl Iterator<Item = hir::Adt> + 'a {
+    // The Option indirection here is to return a single iterator type for both types of field lists.
+    let mut record = None;
+    let mut tuple = None;
+    match list {
+        ast::FieldList::RecordFieldList(list) => {
+            record = Some(
+                list.fields().filter_map(|field| ctx.sema.resolve_type(&field.ty()?)?.as_adt()),
+            );
+        }
+        ast::FieldList::TupleFieldList(list) => {
+            tuple = Some(
+                list.fields().filter_map(|field| ctx.sema.resolve_type(&field.ty()?)?.as_adt()),
+            );
+        }
+    }
+    record.into_iter().flatten().chain(tuple.into_iter().flatten())
+}
+
+type DeriveTarget = (hir::InFileWrapper<HirFileId, ast::Adt>, Vec<DeriveItem>);
+
+fn derive_targets(
+    ctx: &AssistContext<'_>,
+    derives: &[DeriveItem],
+    current_crate: hir::Crate,
+    adt: &hir::Adt,
+) -> Option<DeriveTarget> {
+    // TODO: It might be useful to allow edits to crates within the same workspace.
+    if adt.krate(ctx.db()) != current_crate {
+        return None;
+    }
+
+    let src = ctx.sema.source(*adt)?;
+    let derives: Vec<_> = derives
+        .iter()
+        .filter(|derive| !derive.implemented_by(ctx, adt, &src.value))
+        .cloned()
+        .collect();
+
+    if derives.is_empty() {
+        None
+    } else {
+        Some((src, derives))
+    }
+}
+
+fn field_types_to_derive(
+    ctx: &AssistContext<'_>,
+    ty: &ast::Adt,
+    derives: &[DeriveItem],
+    current_crate: &hir::Crate,
+) -> Option<Vec<DeriveTarget>> {
+    let mut res: Vec<DeriveTarget> = Vec::new();
+    let mut worklist: Vec<_> = match &ty {
+        ast::Adt::Enum(enum_) => enum_
+            .variant_list()?
+            .variants()
+            .filter_map(|variant| variant.field_list())
+            .flat_map(|list| field_list_adts(ctx, list))
+            .filter_map(|adt| Some((adt, derive_targets(ctx, derives, *current_crate, &adt)?)))
+            .collect(),
+        ast::Adt::Struct(struct_) => field_list_adts(ctx, struct_.field_list()?)
+            .filter_map(|adt| Some((adt, derive_targets(ctx, derives, *current_crate, &adt)?)))
+            .collect(),
+        ast::Adt::Union(_) => return None,
+    };
+
+    while let Some((adt, (src, derives))) = worklist.pop() {
+        if res.iter().any(|(res_src, _)| res_src == &src) {
```

Comment:
That's quadratic in the size of the worklist.

(I don't claim this is problematic, but it's something to be aware. Perhaps replace it by a hashset.)